Src/DependencyCollector/Shared/Implementation/TelemetryDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,186 @@
+        internal const string ActivityStopNameSuffix = ".Stop";
+
+        private readonly HashSet<string> excludedDiagnosticSources 
+            = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, HashSet<string>> excludedDiagnosticSourceActivities 
+            = new Dictionary<string, HashSet<string>>(StringComparer.OrdinalIgnoreCase);
+
+
+        public TelemetryDiagnosticSourceListener(TelemetryConfiguration configuration, ICollection<string> excludeDiagnosticSourceActivities) 
+            : base(configuration)
+        {
+            this.Client.Context.GetInternalContext().SdkVersion = SdkVersionUtils.GetSdkVersion("rdd" + RddSource.DiagnosticSourceListener + ":");
+            this.PrepareExclusionLists(excludeDiagnosticSourceActivities);
+        }
+
+        internal override bool IsSourceEnabled(DiagnosticListener value)
+        {
+            return !this.excludedDiagnosticSources.Contains(value.Name);
+        }
+
+        protected override HashSet<string> GetListenerContext(DiagnosticListener diagnosticListener)
+        {
+            HashSet<string> excludedActivities;
+            if (!this.excludedDiagnosticSourceActivities.TryGetValue(diagnosticListener.Name, out excludedActivities))
+            {
+                return null;
+            }
+
+            return excludedActivities;
+        }
+
+        internal override bool IsEventEnabled(string evnt, object input1, object input2, DiagnosticListener diagnosticListener, HashSet<string> context)
+        {
+            return !this.IsActivityExcluded(evnt, context)
+                && !evnt.EndsWith(ActivityStartNameSuffix, StringComparison.OrdinalIgnoreCase);
+        }
+
+        internal override bool ShouldHandleEvent(KeyValuePair<string, object> evnt, DiagnosticListener diagnosticListener, HashSet<string> context)
+        {
+            return !this.IsActivityExcluded(evnt.Key, context)
+                && evnt.Key.EndsWith(ActivityStopNameSuffix, StringComparison.OrdinalIgnoreCase);
+        }
+
+        internal bool IsActivityExcluded(string activityName, HashSet<string> excludedActivities)
+        {
+            return excludedActivities?.Contains(activityName) == true;
+        }
+
+        internal override void HandleEvent(KeyValuePair<string, object> evnt, DiagnosticListener diagnosticListener, HashSet<string> context)
+        {
+            Activity currentActivity = Activity.Current;
+            if (currentActivity == null)
+            {
+                DependencyCollectorEventSource.Log.CurrentActivityIsNull();
+                return;
+            }
+
+            DependencyCollectorEventSource.Log.TelemetryDiagnosticSourceListenerActivityStopped(currentActivity.Id, currentActivity.OperationName);
+
+            // extensibility point - can chain more telemetry extraction methods here
+            ITelemetry telemetry = this.ExtractDependencyTelemetry(diagnosticListener, currentActivity);
```

Comment:
1. some listeners will want to extract telemetry from the `evnt`. 
2. Passing currentActivity is not necessary - not everyone would need it and those who need may use Activity.Current

Reply:
so ```ExtractDependencyTelemetry()``` is just a method to create dependency telemetry from the activity. To try to convert it to sth else new methods would be added here (in a chain-of-responsibility fashion) and those new methods can have different signature depending on their needs

Reply:
what if listener creates Requesttelemetry or OperationTelemetry? 
every virtual method call costs something. Everything that is known about this event is hidden in diagnostics source name, activity name and event payload. Why not have 1 method like: extract telemetry from source, event, payload? What value chain of methods brings?

Reply:
So this particular listener has no fixed extensibility design - it will be figured out when needed. But it is not meant to be subclassed (the ```*Base``` one is), rather more logic should be added in this place - whether in separate methods or the same one that already exists - but that will depend on the actual enhancement that's being made.